sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -135,15 +134,38 @@ secretKeyFromByteString bytestring =
   if BS.length bytestring < fromIntegral cryptoSecretboxKeyBytes
     then Nothing
     else unsafeDupablePerformIO $
-      BS.unsafeUseAsCStringLen bytestring $ \(outsideSecretKeyPtr, _) -> do
-        secretKeyForeignPtr <-
-          BS.mallocByteString @CChar (fromIntegral cryptoSecretboxKeyBytes)
-        Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
-          Foreign.copyArray
-            outsideSecretKeyPtr
-            secretKeyPtr
-            (fromIntegral cryptoSecretboxKeyBytes)
-        pure $ Just $ SecretKey (Foreign.castForeignPtr @CChar @CUChar secretKeyForeignPtr)
+      BS.unsafeUseAsCStringLen bytestring $ \(outsideSecretKeyPtr, _) ->
+        fmap Just $
+          newSecretKeyWith $ \secretKeyPtr ->
+            Foreign.copyArray
+              outsideSecretKeyPtr
+              (Foreign.castPtr @CUChar @CChar secretKeyPtr)
+              (fromIntegral cryptoSecretboxKeyBytes)
+
+-- | Prepare memory for a 'SecretKey' and use the provided action to fill it.
+--
+-- Memory is allocated with 'LibSodium.Bindings.SecureMemory.sodiumMalloc' (see the note attached there).
+-- Finalizer is run when the key is goes out of scope, but 'freeSecretKey' can be used to release early.
+--
+-- @since 0.0.1.0
+newSecretKeyWith :: (Foreign.Ptr CUChar -> IO ()) -> IO SecretKey
+newSecretKeyWith action = do
+  ptr <- sodiumMalloc cryptoSecretboxKeyBytes
+  when (ptr == Foreign.nullPtr) $ do
```

Comment:
I was going to say something about using our foreign pointer equality function (`foreignPtrEq`), then I realized that since it's a null pointer, that wouldn't make sense. Looks fine.